### ISSUE TITLE
Backend broken?

### ISSUE THREAD

**Issue by julienschmidt** [microg/NominatimGeocoderBackend#6]
Lookups on my smartphone fail with `java.io.FileNotFoundException`:

```
04-16 00:48:06.870  7673 14649 D NominatimGeocoderBackend: Requesting http://open.mapquestapi.com/nominatim/v1/reverse?format=json&accept-language=en&lat=39.106944&lon=-76.743056
04-16 00:48:06.981  7673 14649 W NominatimGeocoderBackend: java.io.FileNotFoundException: http://open.mapquestapi.com/nominatim/v1/reverse?format=json&accept-language=en&lat=39.106944&lon=-76.743056
04-16 00:48:06.981  7673 14649 W NominatimGeocoderBackend:  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:238)
04-16 00:48:06.981  7673 14649 W NominatimGeocoderBackend:  at org.microg.nlp.backend.nominatim.BackendService$AsyncGetRequest.run(BackendService.java:189)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend: java.lang.NullPointerException: Attempt to get length of null array
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at java.lang.StringFactory.newStringFromBytes(StringFactory.java:45)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at org.microg.nlp.backend.nominatim.BackendService$AsyncGetRequest.retrieveString(BackendService.java:220)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at org.microg.nlp.backend.nominatim.BackendService.getFromLocation(BackendService.java:57)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at org.microg.nlp.api.GeocoderBackendService$Backend.getFromLocation(GeocoderBackendService.java:72)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at org.microg.nlp.api.GeocoderBackend$Stub.onTransact(GeocoderBackend.java:63)
04-16 00:48:06.981  7673  7690 W NominatimGeocoderBackend:  at android.os.Binder.execTransact(Binder.java:453)
04-16 00:48:06.983 13584 13606 E ReverseGeocoder: Fuuuuu
04-16 00:48:06.983 13584 13606 E ReverseGeocoder: java.io.IOException: no result
04-16 00:48:06.983 13584 13606 E ReverseGeocoder:   at android.location.Geocoder.getFromLocation(Geocoder.java:136)
...
```

If I curl the same URL, the response is `The AppKey submitted with this request is invalid.`. If the backend is not already sending an API key, than this is probably the problem.

**Comment by mar-v-in** [comment]
Duplicate of #4. It is not possible to have an API key shared with a lot of people, so there is no way to publish this backend on F-Droid anymore - without forcing the user to register and enter an API key.

**Comment by julienschmidt** [comment]
Is there any other geocoding backend?

And I think there would be quite a few people who would take the hurdle of registering for an API key. After all we already spend time to get rid of GAPPS.

**Comment by IzzySoft** [comment]
While I basically agree with @julienschmidt I hate to see that one tracker gets replaced by another (same with most weather APIs btw). Though they already can track by the requesting IP, a unique API key makes identification even more unique.

Still: If there was a way to configure an API key, and we knew the limits, multiple users might be able to share a single key privately. As most of us are unlikely to be able to compile the backend from source ourselves: would you mind adding such an option, @mar-v-in ? After all, it will definitely be an improvement. Currently it's not working at all – with a change that lets the user add an own API key, there's at least a choice.

**Comment by ale5000-git** [comment]
@mar-v-in: Please add an option in the GUI to allow the user to change the App Key on the fly (without recompile and without restart the phone).

**Comment by IzzySoft** [comment]
Btw: The [OSM Nominatim Wiki](https://wiki.openstreetmap.org/wiki/Nominatim) states several alternative APIs being available. Next to the [MapQuest Open Initiative](https://wiki.openstreetmap.org/wiki/MapQuest) (which I assume is what's being used here – API key to request [on this site](https://developer.mapquest.com/plan_purchase/steps/business_edition/business_edition_free/register)), there're also
- [PickPoint](http://pickpoint.io/): 2,500 API calls/day free, for non-commercial use. Should fit the average user.
- [OpenCage Geocoder](http://geocoder.opencagedata.com/): free for up to 2,500 requests/day, 1 req/s. Again should fit the average user.
- [LocationIQ](http://locationiq.org/): free developer plan for 30,000 requests / day, 6 req/s, allows even for commercial use.

I haven't checked whether those use API keys to identify users, or if the limits are bound to a requesting IP, or whatever. But it might be a good idea to have some alternative "API servers" available to switch between if one goes broke – and if those are integrated with this backend for that purpose. Of course, implementing them with an alternative backend would be an option as well – so if any dev considers that, please go ahead :)

Other "free reverse geocoding APIs" include:
- [Gisgraphy Gisgraphy](http://www.gisgraphy.com/): also OSM based. Doesn't seem to require an API key at all for its [free service](http://www.gisgraphy.com/free-access.htm). You call the URL and get a JSON object returned. Should be pretty easy to use.
- [Texas A&M Geoservices Reverse Geocoding engine](https://geoservices.tamu.edu/Services/ReverseGeocoding/WebService/v04_01/): free, API key required – but (currently) US only.
- [ArcGIS REST API](https://developers.arcgis.com/rest/geocode/api-reference/geocoding-reverse-geocode.htm): (paid) API key optional, needed if you want to store results in your own database; not needed if you discard results after showing them once.
- [GeoNames Web Services](http://www.geonames.org/export/web-services.html) seems to allow to use a shared key 
- [Mappify.io](https://mappify.io/): 100 req/d per IP without API key, 2,500 r/d with free API key. If a user needs more, throw in cash. Seems nice for a backend: No API key needed to try it, free API key available if one needs more, affordable API key ($1 per 1.000 reqs exceeding the free 2,500) for the few users needing even more after that. Or throw in 1,000$/month for a dedicated server with unlimited requests (kidding – but it's possible).

I'm not an Android dev – but maybe someone with the required knowledge wants to pick some of those and implement them? If done in this backend, hopefully @mar-v-in will merge resulting PR(s)?

**Comment by LuccoJ** [comment]
There is now also an official Nominatim instance from OpenStreetMap at https://nominatim.openstreetmap.org/ and it does not require API keys, while having rate limiting of about a call per second or two.

Generally speaking, since the Nominatim software is open and anyone can potentially set up their own instance (although it does require having a full OSM dataset), it would be ideal if this backend simply provided a setting to change the URL to the user's preferred Nominatim instance, with the official OSM one as the default choice.

**Comment by ArchangeGabriel** [comment]
That would be awesome!

**Comment by Findus23** [comment]
@LuccoJ It seems like the nominatim url is already mentioned in https://github.com/microg/NominatimGeocoderBackend/blob/master/src/main/java/org/microg/nlp/backend/nominatim/BackendService.java#L29
but not used.

**Comment by LuccoJ** [comment]
@Findus23 It also has an extra space before "http". How can I trust @mar-v-in with this unacceptable sloppiness!

**Comment by ale5000-git** [comment]
It may be just a typo, it may happen to everyone.

**Comment by LuccoJ** [comment]
I was just teasing, of course I'm not serious. Anyway it does seem like it should just be a matter of recompiling with the changed URL (though user-configurable would be nicer).

**Comment by ArchangeGabriel** [comment]
This should be fixed by the next build now that #14 landed. ;)
